Remove curandStateMTGP32 usage

syed-ahmed opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#20886 Remove curandStateMTGP32 usage**
* #20626 Speedup bernoulli_scalar_cuda_kernel with grid-stride loop
* #20625 Move THCTensor_(geometric) to ATen
* #20624 Move THCTensor_(lognormal) to ATen
* #20623 Move THCTensor_(exponential) to ATen
* #20622 Move THCTensor_(cauchy) to ATen
* #20621 Move THCTensor_{normal, normal_means, normal_stddevs, normal_means_stddevs} to ATen

Differential Revision: [D15535503](https://our.internmc.facebook.com/intern/diff/D15535503)

## Summary:
This PR removes curandStateMTGP32 usages since it's not stream-safe.
Main changes are:
- It modifies THCTensor_(getRNGState) and THCTensor_(setRNGState) to not read/write curandStateMTGP anymore.
- It modifies RRelu.cu and cuda multinomial kernels to use curandStatePhilox
- It deletes new_state.clone() from torch.cuda.random.py to get a performance boost.

syed-ahmed committed:
Remove curandStateMTGP32 usage

ngimel commented on aten/src/THC/THCTensorRandom.cuh:
```diff
@@ -286,6 +288,10 @@ sampleMultinomialWithReplacement(curandStateMtgp32* state,
   // what, all block threads must participate in the curand_uniform
   // call to update the generator state.
```
This comment is no longer valid (w/o mtgp, individual threads can participate in rng call)

ngimel commented on aten/src/THC/THCTensorRandom.cuh:
```diff
@@ -296,7 +302,8 @@ sampleMultinomialWithReplacement(curandStateMtgp32* state,
       int sample = sampleBase + threadIdx.y;
 
       // All threads participate in this
```
Another invalid comment

ngimel commented on aten/src/THCUNN/generic/RReLU.cu:
```diff
@@ -16,7 +22,12 @@ void THNN_(RReLU_updateOutput)(
+  // each thread will utilize one random, however, since we have to use
+  // curand_uniform4 (See Note [Register spilling in curand call for CUDA < 10]),
+  // offset is 4.
+  uint64_t offset = gen->state.philox_seed_offset.fetch_add(4);
```
note that NUM_BLOCKS for most cases will be set to 64 (that's a poor choice, but for the next PR), so you'll have a grid-stride loop inside the kernel and generate multiple randoms, so adjust offset accordingly.

syed-ahmed replied:
updated offset calc with (numel / block_size * grid.x) * 4.

ngimel reviewed (commented):
Good riddance, mtgp!

syed-ahmed committed:
Update on "Remove curandStateMTGP32 usage"

Remove curandStateMTGP32 usage

gh-metadata: pytorch pytorch 20886 gh/syed-ahmed/8/head

syed-ahmed committed:
Update on "Remove curandStateMTGP32 usage"

Remove curandStateMTGP32 usage

gh-metadata: pytorch pytorch 20886 gh/syed-ahmed/8/head

syed-ahmed committed:
Update on "Remove curandStateMTGP32 usage"

Remove curandStateMTGP32 usage

gh-metadata: pytorch pytorch 20886 gh/syed-ahmed/8/head

ezyang commented:
@syed-ahmed could you rebase this stack on master? (I can do it myself, but if I do you'll have to force update your own local branch pointer.)

syed-ahmed committed:
Update on "Remove curandStateMTGP32 usage"

Remove curandStateMTGP32 usage

gh-metadata: pytorch pytorch 20886 gh/syed-ahmed/8/head

syed-ahmed commented:
@ezyang rebased :).

syed-ahmed committed:
Update on "Remove curandStateMTGP32 usage"

Remove curandStateMTGP32 usage

gh-metadata: pytorch pytorch 20886 gh/syed-ahmed/8/head

ezyang commented:
Sorry, you rebased on top of broken master. Once the breakage is reverted we'll need another rebase :/

ezyang commented:
A little more text in the PR description would have been appreciated for this poor reviewer ^^

ezyang commented on aten/src/THC/THCTensorRandom.cu:
```diff
@@ -47,35 +31,24 @@ THC_API __host__ void THCRandom_getRNGState(THCState* state, THByteTensor *rng_s
   THArgCheck(THByteTensor_nElement(rng_state) == total_size, 1, "RNG state is wrong size");
   THArgCheck(THByteTensor_isContiguous(rng_state), 1, "RNG state must be contiguous");
-  THCudaCheck(cudaMemcpy(THByteTensor_data(rng_state), gen->state.gen_states,
-                         states_size, cudaMemcpyDeviceToHost));
```
It might be a good idea to fill in this memory with deterministic garbage so if someone tries to use it (improperly) it won't be a random error

syed-ahmed replied:
Filled in the memory with -1 and verified locally that torch.cuda.get_rng_state() gives 255 in the first few elements.

ezyang commented on aten/src/THC/THCTensorRandom.cu:
```diff
@@ -87,11 +60,6 @@ THC_API __host__ void THCRandom_setRNGState(THCState* state, THByteTensor *rng_s
   THArgCheck(THByteTensor_isContiguous(rng_state), 1, "RNG state must be contiguous");
-
-  THCudaCheck(cudaMemcpy(gen->state.gen_states, THByteTensor_data(rng_state),
-                         states_size, cudaMemcpyHostToDevice));
```
Ditto here

syed-ahmed replied:
Is this necessary? Since I made all the gen_states memory to have -1 in `getRNGState`, this function will just not affect that value. If I were to do cudaMemcpy or memset here, I need to allocate the gen_states (which I deleted i.e. the initializeGenerator function).

ezyang commented on aten/src/THC/THCTensorRandom.cuh:
```diff
@@ -273,7 +275,7 @@ sampleMultinomialOnce(int64_t* dest,
 template <typename T>
 __global__ void
-sampleMultinomialWithReplacement(curandStateMtgp32* state,
+sampleMultinomialWithReplacement(std::pair<uint64_t, uint64_t> seeds,
```
To be fair, the second element of this pair isn't really a seed, it's an offset, right?

syed-ahmed replied:
That's true. Little hand wavy here I agree. But you could interpret it as, since seed decides where a rng sequence starts from, the offset just gives a finer control over it for the philox engine. So seed for philox could be an umbrella term for the actual seed value + offset :man_shrugging: . If you want I can change the name (`seed_and_offset` maybe?), but then we should be changing the variable name every where and use it like `seed_and_offset.first`, `seed_and_offset.second`.

ezyang replied:
OK, if you like it, let's keep it :)

ezyang commented:
How can I tell if the offset calculations were done right? Do tests cover this at all? It seems very fiddly.

syed-ahmed commented:
> How can I tell if the offset calculations were done right? Do tests cover this at all? It seems very fiddly.

The philox offset calculation for RRelu.cu should be good, since it runs the exact same way as the kernels tested in `cuda_distributions_test.cu`.  I can add a test for multinomial.

syed-ahmed committed:
Update on "Remove curandStateMTGP32 usage"

Remove curandStateMTGP32 usage

gh-metadata: pytorch pytorch 20886 gh/syed-ahmed/8/head

syed-ahmed committed:
Update on "Remove curandStateMTGP32 usage"

Remove curandStateMTGP32 usage

gh-metadata: pytorch pytorch 20886 gh/syed-ahmed/8/head

syed-ahmed committed:
Update on "Remove curandStateMTGP32 usage"

Remove curandStateMTGP32 usage

gh-metadata: pytorch pytorch 20886 gh/syed-ahmed/8/head

ezyang commented:
Sorry, we need another rebase; master was a disaster yesterday.

syed-ahmed committed:
Update on "Remove curandStateMTGP32 usage"

Remove curandStateMTGP32 usage

gh-metadata: pytorch pytorch 20886 gh/syed-ahmed/8/head

syed-ahmed committed:
Update on "Remove curandStateMTGP32 usage"

Remove curandStateMTGP32 usage

gh-metadata: pytorch pytorch 20886 gh/syed-ahmed/8/head

syed-ahmed committed:
Update on "Remove curandStateMTGP32 usage"

Remove curandStateMTGP32 usage

gh-metadata: pytorch pytorch 20886 gh/syed-ahmed/8/head

syed-ahmed committed:
Update on "Remove curandStateMTGP32 usage"

Remove curandStateMTGP32 usage

gh-metadata: pytorch pytorch 20886 gh/syed-ahmed/8/head

syed-ahmed committed:
Update on "Remove curandStateMTGP32 usage"

Remove curandStateMTGP32 usage

gh-metadata: pytorch pytorch 20886 gh/syed-ahmed/8/head

syed-ahmed committed:
Update on "Remove curandStateMTGP32 usage"

Remove curandStateMTGP32 usage

gh-metadata: pytorch pytorch 20886 gh/syed-ahmed/8/head
